### PR TITLE
Add some tests and Travis integration

### PR DESCRIPTION
**Pull request by Peque**
These would be the results as reported by Travis:

https://travis-ci.org/Peque/mazefiles/builds/356025295

I am currently checking basic stuff about the maze file formatting and correctness. I also check that the file name only contains characters in "a-z", "A-Z", "0-9", "-" and "_". I could modify that to allow only lower case in names and either "-" or "_". Although that would require some name changes if we wanted the tests to pass, of course. :blush:

By the way, in order to enable Travis in your `mazefiles` repository you need to activate that service. It is free and you can register using your GitHub account. I can do that if you want, but I would require admin access to the repository.

### CODE REVIEW

**Comment by micromouseonline** [comment]
I have enable travis integration on the mazefiles and the maze-tool repos.

As far as I can see, all the names should satisfy the [a..z0..9\-\.] requirement. I think that uppercase characters should be disallowed in maze filename because of the ambiguous treatment across filesystems.
In the root, files like README should be uppercase if desired but not enforced.


Peter Harrison
peter.harrison@helicron.net

**Comment by Peque** [comment]
Okay, I will change that. Some tests will fail until all names satisfy the requirements, but I guess that is fine for now.

I will also add a half-size example, as I have realized I removed the `halfsize` directory when deleting the `.gitignore` file inside it. I think I have a working basic maze image to maze text file script, which should help creating a collection of half-sized mazes (it would really be a pain to make those manually... :joy:).

**Comment by Peque** [comment]
BTW: I will add some commits as "fixup" to previous commits, so when you think this is ready to be merged, tell me and I will first squash them.

**Comment by micromouseonline** [comment]
git cannot track empty folders. That is why the empty .gitignore was there. As long as there is something present, it will be fine.

I do not know of anyone with halfsize maze files. I will ask around.

Peter Harrison
peter.harrison@helicron.net

**Comment by Peque** [comment]
Wooops, I just changed the `_` in file names too (there were only 3 occurrences, so I thought it would be better to have more homogeneous naming and just use `-` everywhere). I can revert that if you want.

Yeah, I will update the README file too! Good idea to document this a bit.

**Comment by Peque** [comment]
I do not think that is necessary. After all, the word "classic" and "halfsize" is already in the path as their respective parent directories. The Python script already checks for size consistency, according to the path (i.e.: 16x16 for files under "classic", 32x32 for files under "halfsize").

By the way, the half-size example I uploaded I found it here:

http://www.ntf.or.jp/mouse/micromouse2014/MM2014recode01.html

There are some other mazes from previous years, but it is definitely far from a complete collection. Only good news is that the script that I wrote seems to work fine with those images! (unless I missed something, the generated text file seems to be correct) :smile:

**Comment by micromouseonline** [comment]
OK. We will go with it as-is for now.

You sucked that data out of the image? Good job!

Peter Harrison
peter.harrison@helicron.net

**Comment by Peque** [comment]
Updated the README file. Took the liberty of adding a "FAQ" section. Do you like it?

About the image->text script yeah, it worked surprisingly well and was surprisingly easy (the current, very basic implementation is 60 lines of code). Still need to try it with not-so-clean images to see what happens. :joy: Will share it soon too!

About marking the goal cell at (26, 5), I did not know that. Why is it important? Is not always the only section the mouse can reach that is open just like in the classic competition? (i.e.: no walls in that square/rectangle) Also, is that information provided to the contestants before the competition? As you can see, I am not very well informed about half-size competitions... :sweat_smile:

Anyway, I will change that tomorrow! :blush:

**Comment by micromouseonline** [comment]
Marking the goal is only needed for the halfsize contest. The goal can be any single cell and its location is given to the contestants well in advance of the contest.

Since it can be any cell, there is no reason that it should not be halfway along a corridor or a single cell at the end of a cul-de-sac.

The contestant cannot rely on an open area.

Peter Harrison
peter.harrison@helicron.net

**Comment by Peque** [comment]
Okay, I will add the goal then. :blush:

In the README you say goal cells are marked with an asterisk. I like a "G" more. Would you agree in using the letter "G" instead?

We could also, for completeness, allow an "S" at the start cell, even if it is always at (0, 0), it could be useful for training/non-standard mazes.

**Comment by micromouseonline** [comment]
Sure. That makes it match the Japanese mazes.

**Comment by Peque** [comment]
Pushed those changes. Tell me if there is anything else you want me to change.

If everything looks good to you I will squash all the fixups before you merge. :+1:

**Comment by Peque** [comment]
Hmmmm... I was thinking maybe we could be less restrictive. In example:

- We specify, as we do now, that the text maze format allows (but does not enforce) the use of "G" to mark goal cell(s).
- We can specify in the README that halfsize *must* have a goal cell marked, though (I added that check to the tests). This is to comply with the international rules. I understand that "classic" and "halfsize" directories should comply with international rules.
- But even classic mazes, they could have the goal cells (the four cells) marked with a "G", just like here: http://www.ntf.or.jp/mouse/micromouse2014/MM2014recode01.html
- Maybe we could even force all of the mazes to have the goal cell(s) marked with a "G". All classic mazes should have the four center cells marked with a G in that case. I think this could be a good idea.
- This allows for training mazes, for example, to have a goal cell marked anywhere. Or multiple goal cells which might even be separated. After all, training mazes do not have to comply with any international rules.

So at the end: the text maze format only specifies that goal cells are to be marked with a "G". Then, depending of the rules (if any), we apply that specification to all the maze files in the different directories.

What do you think?

**Comment by micromouseonline** [comment]
In classic contest mazes, the goal is implicit. In the halfsize maze, it is explicit.
Thus, a contest halfsize MUST have a goal marked but no others need it. Without that, the record represented by a halfsize contest maze file is useless.
Training mazes implicitly have no specific goal. If they encompass the classic goal, it can be detected. If not, it is entirely at the discretion of the user whether the layout is classic or halfsize. 

Peter Harrison
peter.harrison@helicron.net

**Comment by micromouseonline** [comment]
. . . And we could move any non-contest mazes, regardless of size, into training.

Peter Harrison
peter.harrison@helicron.net

**Comment by Peque** [comment]
Yeah, agree about moving non-contest mazes to training. That should be a separate pull-request though.

I pushed the changes you requested in the README file too.

Just to try to explain myself better one last time (I think maybe I did not the first time)... What I had in mind is:

- In the list that defines the file format in the README file, we only mention that a "G" can be written to mark goal cell(s).
- Bellow, when talking about the `halfsize` folder, we clearly specify that the "G" is a must and there is only one goal cell. This folder is for mazes that comply with the international halfsize rules.
- In training mazes we could allow any configuration, not necessarily complying with classic nor halfsize international maze rules. Although that could be in a different folder... Here we could have a 50x50 maze, for example, or a 4x6 not contained in a 16x16 maze. Goal cells could be single, groups or even multiple and separated. No rules except to comply with the maze file format. 🤪

But I understand if you want to exclude that from your collection.

**Comment by Peque** [comment]
Actually, instead of moving non-contest mazes to `training`, we could prepend the word `training` to their names. This allows for easy filtering. While we still leave totally valid classic maze configurations in the same folder.

**Comment by Peque** [comment]
By the way, I just finished packaging the software for image-to-maze-file conversion:

https://github.com/Theseus/ommr

Is it fine for you if I mention it in your README? Maybe potential contributors could benefit from it making their job a bit simpler.

If you have some spare time and want to try it out I would appreciate some feedback! :blush: All the documentation (very short) is in the README file of the repository.

**Comment by micromouseonline** [comment]
That should work

**Comment by micromouseonline** [comment]
Yes. Include the link.

**Comment by micromouseonline** [comment]
That works really well.

I note that the default seems to be to assume halfsize rather than classic.

If it could cope with not having the right/bottom walls cropped, it would be possible to scrape images from web pages/sites and auto-convert them.


Peter Harrison
peter.harrison@helicron.net

**Comment by micromouseonline** [comment]
Excuse my ignorance but - how does pip know where to find ommr?

**Comment by Peque** [comment]
> Yes. Include the link.

I just pushed the update in the README file. Tell me if there are any other changes you want me to do before squashing this.

> If it could cope with not having the right/bottom walls cropped, it would be possible to scrape images from web pages/sites and auto-convert them.

Yeah, I already [opened an issue for that](https://github.com/Theseus/ommr/issues/7). :blush:

Are there any websites in particular you are thinking about scraping? I thought I could create a simple script to crop all images in [http://www.ntf.or.jp/mouse/history/index.html](http://www.ntf.or.jp/mouse/history/index.html) and that should work well but then: it would only work for those images. Matching names between those images and your repo was another task that required some extra work. Would you help me with that last part if I did the first? :wink:

> Excuse my ignorance but - how does pip know where to find ommr?

I uploaded the package to the public [PyPI repository](https://pypi.python.org/pypi). Once you have your `setup.py` file configured it is really simple. :blush:

**Comment by micromouseonline** [comment]
No sites in particular spring to mind. Sometimes I find them on Japanese sites. If it were possible to just take an image link and feed it to the converter, that would be super convenient. For the lazy like me that is.

The NTF archive looks handy. It would be good to do a bulk conversion and compare with what we have. I can help with name reconciliation where I am reasonably sure I can work it out.

The names in my repos come from all over and are pretty cryptic in many caes. I would love to know what event some of the mazes actually represent.

**Comment by Peque** [comment]
About the cryptic names... I would prefer longer, explicit names like `japan-2011-expert-final` instead of `japan2011ef`. It is easier to search for "final" than "f" (would give better results). And if you want to display shorter names, removing dashes and cropping words is always easier than the reverse operation...

But I guess you do prefer shorter names, right? :joy:

**Comment by micromouseonline** [comment]
No. Explicit names are much better. Most of the originals for these files probably come from the days of 8.3 filenames and then they just got made more consistent.

I have no problem with longer names if they are well defined and consistent. Some of the Japanese names form their archive are a bit wordy though.

I will try and look through the pull request later. Busy couple of days here.

**Comment by Peque** [comment]
Had to search what 8.3 filenames were. :joy:

Great! No hurries, of course. :+1:

**Comment by Peque** [comment]
Squashed everything into 4 commits to hopefully make the final review a bit easier.
